File: Categories/Input/Drag/raw_input.go
package Drag

type RawInputMsg struct {
	Kind       string
	X          float64
	Y          float64
	RectLeft   float64
	RectTop    float64
	RectWidth  float64
	RectHeight float64
	Button     int
	Ctrl       bool
	Shift      bool
	Alt        bool
	Meta       bool
	DeltaX     float64
	DeltaY     float64

	Hit RawHit

	Key string
}

type RawHit struct {
	Kind string

	PortRow int

	EdgeRow int

	NodeRow int
	IsInput bool
}

func DecodeRawInput(rec []byte) (RawInputMsg, bool) {
	if len(rec) == 0 || rec[0] != KindRawInput {
		return RawInputMsg{}, false
	}
	return decodeRawInputFrom(NewReader(rec, 1))
}

File: Categories/Scene/Drag/raw-input.ts
export type RawPointerKind =
  | "pointerdown"
  | "pointermove"
  | "pointerup"
  | "wheel"
  | "home"
  | "delete"
  | "key";

export type RawHit = {
  kind: "port" | "handhold" | "node" | "edge" | "torus" | "empty";
  isInput: boolean;

  onRim: boolean;

  nodeRow: number;
  portRow: number;

  edgeRow: number;

  pointX: number;
  pointY: number;
  pointZ: number;
};

export type RawInputEvent = {
  kind: RawPointerKind;
  x: number;
  y: number;
  rectLeft: number;
  rectTop: number;
  rectWidth: number;
  rectHeight: number;
  button: number;
  ctrl: boolean;
  shift: boolean;
  alt: boolean;
  meta: boolean;
  deltaX: number;
  deltaY: number;
  hit: RawHit;
  key?: string;

  ballX: number;
  ballY: number;
  ballZ: number;

  ballPrevX: number;
  ballPrevY: number;
  ballPrevZ: number;
};
